### PR TITLE
Chrome extension

### PR DESCRIPTION
**Pull request by Malpp**
Merge the chrome extension + code review

### CODE REVIEW

File: misc/chrome_extension/content-script.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
+'use strict';
+//The global timeout variable for the addToPage()
+var addTimeOut;
+var apiKey = 'xx0b957ee7-8846-4b6c-b4c3-6f88362e601f';
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: Peut-être ajouter un commentaire pour dire qu'il s'agit d'une clé d'API publique?

File: misc/chrome_extension/content-script.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+						setAceEditorValue('');
+						$('#BodyTextDataStream').attr('checked', false);
+						$('#BodyHTMLDataStream').attr('checked', false);
+						$('#ThumbnailDataStream').attr('checked', false);
+						$('#FileBinaryStream').attr('checked', false);
+						$('#ExtensionName').val('');
+						$('#ExtensionDescription').val('');
+
+						if (uniqueId) {
+							$.get(`https://platformqa.cloud.coveo.com/rest/search/v2/html?organizationId=extensions&uniqueId=${uniqueId}&access_token=${apiKey}`,
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: Je suggérerais d'avoir les valeurs comme les url cibles dans un fichier à part représentant la configuration, et ensuite de référence cette config là dans le code.

File: misc/chrome_extension/content-script.js (hunk 1 above)
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: Aussi, centraliser les éléments de configuraiton (voir commentaire suivant).

File: misc/chrome_extension/content-script.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	//Get the HTML data
+	$.get(chrome.extension.getURL('/modal.html'), function (data) {
+		let containerDiv = document.createElement('div');
+		containerDiv.innerHTML = data;
+		editorElement.insertBefore(containerDiv, editorElement.childNodes[0]);
+
+		//Init the Coveo search
+		var root = document.getElementById('__search');
+		Coveo.SearchEndpoint.endpoints['default'] = new Coveo.SearchEndpoint({
+			restUri: 'https://platformqa.cloud.coveo.com/rest/search',
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: Centraliser la configuration (voir commentaire précédent)

File: misc/chrome_extension/content-script.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+			//Public key with only search enabled
+			//accessToken: 'xx55c20bcb-59aa-40a2-b8b2-72ae625e6762'
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: Enlever le commentaire si la clé est spécifiée plus haut.

File: misc/chrome_extension/content-script.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+			createModal();
+
+		}
+	}, 350);
+}
+
+
+window.onload = function () {
+	// your code
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: ?

File: misc/chrome_extension/content-script.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+	//Checks if there were changes 
+	MutationObserver = window.MutationObserver || window.WebKitMutationObserver;
+
+	var observer = new MutationObserver(function (mutations, observer) {
+		// fired when a mutation occurs
+		if (document.getElementById('EditExtensionComponent')) {
+			addToPage();
+		}
+		// ...
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: ?

File: misc/chrome_extension/content-script.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+	});
+
+	// define what element should be observed by the observer
+	// and what types of mutations trigger the callback
+	observer.observe(document, {
+		subtree: true,
+		attributes: true
+	});
+
+	//Output of seachAPI
```
Comment (ancientwinds) [inline misc/chrome_extension/content-script.js]: À enlever.

**Comment by Malpp** [comment]
@ancientwinds I added an options page inside the extension
image

Those are the default values

**Comment by Malpp** [comment]
After working on it the past 3 days, the extension tester is now in a very good position. The last commit I just did should make it able to test all the documents on a source no matter what, although I would need some sort of unit test to confirm that.

**Comment by Malpp** [comment]
This one was a hard one to solve, god.
But it now checks if an extension requested either a "Body Text" or "Body HTML" and gives it to the extension. So more extensions can be tested

File: misc/chrome_extension/README.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+### The gallery
+The extension gallery can be found when one clicks on the "Add Extension" button in the top right of the extensions page.
+This is what is looks like:
+![extension gallery button](https://user-images.githubusercontent.com/17149559/27692683-569b0ad6-5cb5-11e7-996f-1d4f8e12c08e.png)
+When clicked on, a Coveo seach pops up with a list of availible extensions. The extensions are added by the python script portion of the repo.
+![extension gallery search](https://user-images.githubusercontent.com/17149559/27692821-abeb7b88-5cb5-11e7-8b81-17793b94eb6f.png)
+
+#### Additional notes
+If the search does not work, the org may have changed place in the future/does not exist anymore.
+It is currently living in the CoveoStageing in platformqa.
```
Comment (jdevost) [inline misc/chrome_extension/README.md]: CoveoStageing --> Coveo's staging platform.

I wonder if it should stay in platformqa. We should confirm with JF.
I think (well, I hope) it's only available to Coveo people.

File: misc/chrome_extension/content-script.js [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+	}
+
+
+	/**
+	 * Add an error message to the test
+	 * 
+	 * @param {string} str - The error message
+	 */
+	function addError(str) {
+		let message =
```
Comment (jdevost) [inline misc/chrome_extension/content-script.js]: addError and addWarning are the same, but for the class. 
Duplicated code often leads to bugs where one function is updated, and not the other. 

Can we merge them: 
```/* default to error*/
function addMessage(msg, msgType) {
  let message = `<div class='banner flex center-align bg-${msgType=='warning'?'yellow':'red'}'>
    <div class="banner-description">
       <p>${msg}</p>
    </div></div>`;
  errorBannerElement.append(message);
}
```

File: misc/chrome_extension/content-script.js [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+ * The 'init' function of the script
+ * Loads the values from the config and inits the mutation obs
+ * 
+ */
+window.onload = function () {
+
+	//Default values if no values are found
+	chrome.storage.local.get({
+		// Public key with only search enabled
+		__publicApiKey: 'xx0b957ee7-8846-4b6c-b4c3-6f88362e601f',
```
Comment (jdevost) [inline misc/chrome_extension/content-script.js]: Can these values (api key, platform) be in a config file ?

File: misc/chrome_extension/content-script.js [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+	script.appendChild(document.createTextNode(scriptContent));
+	(document.body || document.head || document.documentElement).appendChild(script);
+
+	$('#tmpScript').remove();
+
+}
+
+
+/**
+ * Converts a UTF-8 string into UTF-16LE
```
Comment (jdevost) [inline misc/chrome_extension/content-script.js]: Can you describe here why we need to do that conversion? 
It's not obvious to a new reader (and we will forget too)

File: misc/chrome_extension/js/options.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	});
+}
+
+// Restores select box and checkbox state using the preferences
+// stored in chrome.storage.
+function restore_options() {
+	// Use default value
+	chrome.storage.local.get({
+		//Public key with only search enabled
+		__publicApiKey: 'xx0b957ee7-8846-4b6c-b4c3-6f88362e601f',
```
Comment (jdevost) [inline misc/chrome_extension/js/options.js]: Again, can this be moved in a config file? Or in the `manifest.json` maybe?

File: misc/chrome_extension/README.md (hunk 8 above)
Comment (Malpp) [inline misc/chrome_extension/README.md]: I already tried asking him but he never gave me a clear answer. It was in platformdev before, so at least we're doing progress.

File: misc/chrome_extension/content-script.js (hunk 11 above)
Comment (Malpp) [inline misc/chrome_extension/content-script.js]: Added lots of comments to it in the lastest commit

**Pull request merged**